Sources/Customization/BatchEventProcessor.swift

```diff
@@ -0,0 +1,271 @@
+        guard dataStore.count < maxQueueSize else {
+            let error = OptimizelyError.eventDispatchFailed("EventQueue is full")
+            self.logger.e(error)
+            completionHandler?(.failure(error))
+            return
+        }
+        
+        guard let body = try? JSONEncoder().encode(event) else {
+            let error = OptimizelyError.eventDispatchFailed("Event serialization failed")
+            self.logger.e(error)
+            completionHandler?(.failure(error))
+            return
+        }
+        
+        dataStore.save(item: EventForDispatch(body: body))
+        
+        if dataStore.count >= batchSize {
+            flush()
+        } else {
+            startTimer()
+        }
+        
+        completionHandler?(.success(body))
+    }
+
+    // notify group used to ensure that the sendEvent is synchronous.
+    // used in flushEvents
+    let notify = DispatchGroup()
+    
+    open func flush() {
+        dispatcher.async {
+            // we don't remove anthing off of the queue unless it is successfully sent.
+            var failureCount = 0
+            
+            func removeStoredEvents(num: Int) {
+                if let removedItem = self.dataStore.removeFirstItems(count: num), removedItem.count > 0 {
+                    // avoid event-log-message preparation overheads with closure-logging
+                    self.logger.d({ "Removed stored \(num) events starting with \(removedItem.first!)" })
+                } else {
+                    self.logger.e("Failed to removed \(num) events")
+                }
+            }
+            
+            while let eventsToSend: [EventForDispatch] = self.dataStore.getFirstItems(count: self.batchSize) {
+                let (numEvents, batched) = eventsToSend.batch()
+                
+                guard numEvents > 0 else { break }
+                
+                guard let batchEvent = batched else {
+                    // discard an invalid event that causes batching failure
+                    // - if an invalid event is found while batching, it batches all the valid ones before the invalid one and sends it out.
+                    // - when trying to batch next, it finds the invalid one at the header. It discards that specific invalid one and continue batching next ones.
+
+                    removeStoredEvents(num: 1)
+                    continue
+                }
+                
+                // we've exhuasted our failure count.  Give up and try the next time a event
+                // is queued or someone calls flush.
+                if failureCount > DefaultValues.maxFailureCount {
+                    self.logger.e(.eventSendRetyFailed(failureCount))
+                    break
+                }
+                
+                // send notification BEFORE sending event to the server
+                NotificationCenter.default.post(name: .willSendOptimizelyEvents, object: batchEvent)
```

Comment:
Will take a look at it. An issue is that we do not have sdkKey to inject the right NotificationCenter since EventProcessor is global, shared for all sdkKeys.

Reply:
That is a good point.  But, this breaks down as well.  What if you have multiple clients running?  Wouldn't they all be notified?  In general the notification center is not a great mechanism when working on a large team.  In previous positions, we discovered that debugging became much more difficult using notification center.  Support for multiple delegates to be notified worked much better and was clearer when trying to find delegate implementations.